analyzers/src/SonarAnalyzer.Common/Common/DisjointSets.cs

```diff
@@ -0,0 +1,47 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Common;
+
+/// <summary>
+/// Data structure for working with disjoint sets of strings, to perform union-find operations with equality semantics:
+/// i.e. reflexivity, symmetry and transitivity. See https://en.wikipedia.org/wiki/Disjoint-set_data_structure.
+///
+/// An example of use is to build undirected connected components of dependencies, where each node is the identifier.
+///
+/// Uses a dictionary of strings as a backing store. The dictionary represents a forest of trees, where each node is
+/// a string and each tree is a set of nodes.
+/// </summary>
+public class DisjointSets
+{
+    private readonly Dictionary<string, string> parents;
+
+    public DisjointSets(IEnumerable<string> elements) =>
+        parents = elements.ToDictionary(x => x, x => x);
+
+    public void Union(string from, string to) =>
+        parents[FindRoot(from)] = FindRoot(to);
+
+    public string FindRoot(string element) =>
+        parents[element] is var root && root != element ? FindRoot(root) : root;
+
+    public List<List<string>> GetAllSets() =>
```

Comment:
I prefer it as a method, since "property" semantically means light-weight execution, and this method has a lot of logic.

On a more general note, what about not exposing any of the `Union`, `FindRoot` methods?
Isn't what we want from this class always the same?

We could make the class static and expose only an `Execute`, and return the list of lists. E.g.:

```csharp
var groups = DisjointSets.Execute(myList);
```

What is the reason that the caller needs to care about the algorithm's implementation?
This is even more true now that we renamed it from `DisjointSetPrimitives` to `DisjointSets` and made it non-static.

Reply:
I like @gregory-paidis-sonarsource suggestion even more.

Reply:
We discussed it offline, here is the wrap-up:

- It was mentioned that maybe we should have left this functionality in the rule, as extracting it prematurely as a stand-alone data structure does not have that much value.
- For now, to be efficient, let's keep this as it is and focus on merging the rule to peach validate.

(Antonio feel free to resolve this when you read it)

Reply:
I would expect the person who opened the discussion thread to resolve it.
@mary-georgiou-sonarsource As a reviewer and initiator of the proposal to extract the three methods into a dedicated PR, please solve the discussion if you agree with the conclusion above.